src/dials/algorithms/correlation/analysis.py

```diff
@@ -0,0 +1,211 @@
+        self.identifiers_to_ids_map = {
+            value: key for key, value in self.ids_to_identifiers_map.items()
+        }
+
+        if "inverse_scale_factor" in reflections[0]:
+            # FIXME - can we also just use the filtered_arrays_from_experiments_reflections function?
+            from dials.report.analysis import scaled_data_as_miller_array
+
+            datasets = []
+            for expt, r in zip(self._experiments, self._reflections):
+                sel = ~r.get_flags(r.flags.bad_for_scaling, all=False)
+                sel &= r["inverse_scale_factor"] > 0
+                datasets.append(scaled_data_as_miller_array([r], [expt]))
+
+        else:
+            datasets = filtered_arrays_from_experiments_reflections(
+                self._experiments,
+                self._reflections,
+                outlier_rejection_after_filter=False,
+                partiality_threshold=params.partiality_threshold,
+            )
+        individual_merged_intensities = []
+        for unmerged in datasets:
+            individual_merged_intensities.append(
+                unmerged.merge_equivalents().array().set_info(unmerged.info())
+            )
+        self.datasets = [
+            d.eliminate_sys_absent(integral_only=True).primitive_setting()
+            for d in individual_merged_intensities
+        ]
+
+        self.params.lattice_group = self.datasets[0].space_group_info()
+        self.params.space_group = self.datasets[0].space_group_info()
+
+        self.cosym_analysis = CosymAnalysis(self.datasets, self.params)
+
+    def _filter_min_reflections(self, experiments, reflections):
+        """
+        Filter all datasets that have less than the specified number of reflections.
+        """
+        identifiers = []
+
+        for expt, refl in zip(experiments, reflections):
+            if len(refl) >= self.params.min_reflections:
+                identifiers.append(expt.identifier)
+
+        return select_datasets_on_identifiers(
+            experiments, reflections, use_datasets=identifiers
+        )
+
+    def calculate_matrices(self):
+        """
+        Calculates the cc and cos angle correlation matrices for the input dataset.
+        The cc matrix is converted to a distance matrix, and both are condensed into a nC2 array.
+        Matrices are converted to plotly json files for visual output.
+        """
+
+        self.cosym_analysis._intialise_target()
+        self.cosym_analysis._determine_dimensions()
+        self.cosym_analysis._optimise(
+            self.cosym_analysis.params.minimization.engine,
+            max_iterations=self.cosym_analysis.params.minimization.max_iterations,
+            max_calls=self.cosym_analysis.params.minimization.max_calls,
+        )
+
+        logger.info("\nCalculating Correlation Matrix (rij matrix - see dials.cosym)\n")
+
+        correlation_matrix = self.cosym_analysis.target.rij_matrix
+
+        for i in range(correlation_matrix.shape[0]):
```

Comment:
I think it would be better to separate out L140 - L160 into two separate functions, so that they can be called instead of `compute_correlation_coefficient_matrix` and `compute_cos_angle_matrix` in xia2/Modules/MultiCrystal/__init__.py

Reply:
Agreed, that does make future integration into multiplex easier :) Addressed in commit f4e4de1

Reply:
If we wanted to replace the calls in the current multiplex implementation, I think we need either standalone functions or static methods that have the input arguments of rij_matrix and the cosym.coords respectively? We wouldn't want to be instantiating the CorrelationMatrix class in multiplex as the input there is not experiments and reflections but miller arrays.

Reply:
True! They're both static methods now for flexibility :)